src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ObjectiveCMarshal.NativeAot.cs

```diff
@@ -51,20 +55,130 @@ internal static void ThrowPendingExceptionObject()
             }
         }
 
+        [RuntimeExport("ObjectiveCMarshalTryGetTaggedMemory")]
+        static bool TryGetTaggedMemory(IntPtr pObj, IntPtr* tagged)
+        {
+            // We are paused in the GC, so this is safe.
+            object obj = Unsafe.AsRef<object>((void*)&pObj);
+
+            if (!s_objects.TryGetValue(obj, out ObjcTrackingInformation? info))
+            {
+                return false;
+            }
+
+            *tagged = info._memory;
+            return true;
+        }
+
+        [RuntimeExport("ObjectiveCMarshalGetIsTrackedReferenceCallback")]
+        static IntPtr GetIsTrackedReferenceCallback()
+        {
+            return s_IsTrackedReferenceCallback;
+        }
+
+        [RuntimeExport("ObjectiveCMarshalGetOnEnteredFinalizerQueueCallback")]
+        static IntPtr GetOnEnteredFinalizerQueueCallback()
+        {
+            return s_OnEnteredFinalizerQueueCallback;
+        }
+
         private static bool TryInitializeReferenceTracker(
             delegate* unmanaged<void> beginEndCallback,
             delegate* unmanaged<IntPtr, int> isReferencedCallback,
             delegate* unmanaged<IntPtr, void> trackedObjectEnteredFinalization)
         {
-            throw new NotImplementedException();
+            if (Interlocked.CompareExchange(ref s_initialized, 1, 0) != 0)
+            {
+                return false;
+            }
+
+            if (!RuntimeImports.RhRegisterObjectiveCMarshalBeginEndCallback((IntPtr)beginEndCallback))
+            {
+                // We failed to allocate unmanaged memory, so no state has been changed.
+                // Reset s_initialized so initialization can be attempted again.
+                s_initialized = 0;
+                throw new OutOfMemoryException();
+            }
```

Comment:
This failure case is unnecessary because only one set of callbacks is supported at a time. This relates to the linked list feature that has been added here.

Reply:
There is still a failure case here after removing the linked list, in the multiple managed worlds situation discussed in other parts of this PR.

Currently I have it failing relatively gracefully: an exception will be thrown saying the Objective-C marshaling was initialized twice. Perhaps it should be a fail fast instead?

Reply:
I simplified this further by removing the Interlocked.CompareExchange in managed code and instead relying on the callback registration in native code to ensure exactly-once initialization.